File: runner_scripts/resume.py
"""Continuation of partial runs against the result store; mirrored by resume.jl.

BENCH_RESUME=1 skips every point with a row in the store (NaN rows count).
BENCH_NO_OVERWRITE=1 skips only points whose row holds a finite time.
BENCH_RESUME_FROM is a cursor problem[:algorithm][:fixed|adaptive][:N] into the
run order (problems.csv, then algorithms.csv, fixed before adaptive, N
ascending); points strictly before it are skipped. The problem[:N] form floors
every leg of that problem at N; in the states sweep N is the state count. A wp
leg is skipped only when every setting is covered. BENCH_FLOOR=1 skips nothing
and each result merges into the store keeping the lower time.
"""

import os

from algorithms import MODES, algorithm_names, get_algorithm
from problems import get_problem, problem_names
from protocol import N_WP, STATES_N, TOLS
from results import floor_enabled  # noqa: F401

_CURSOR_CACHE = []          # [] = unparsed, [None] or [dict] once parsed


def resume_enabled():
    """True when BENCH_RESUME asks for skip-what-is-on-disk continuation."""
    return os.environ.get("BENCH_RESUME", "") not in ("", "0")


def no_overwrite_enabled():
    """True when BENCH_NO_OVERWRITE asks to keep only finite recorded rows."""
    return os.environ.get("BENCH_NO_OVERWRITE", "") not in ("", "0")


def parse_cursor(spec):
    """BENCH_RESUME_FROM spec -> cursor dict; omitted parts are None."""
    parts = [tok for tok in spec.split(":")]
    if not parts or not parts[0]:
        raise SystemExit("BENCH_RESUME_FROM requires a problem name, got "
                         "'{0}'".format(spec))
    get_problem(parts[0])
    cursor = {"problem": problem_names().index(parts[0]),
              "algorithm": None, "mode": None, "n": None}
    for tok in parts[1:]:
        if tok.isdigit():
            if cursor["n"] is not None:
                raise SystemExit("BENCH_RESUME_FROM '{0}': more than one N"
                                 .format(spec))
            cursor["n"] = int(tok)
        elif tok in MODES:
            if cursor["algorithm"] is None or cursor["mode"] is not None \
                    or cursor["n"] is not None:
                raise SystemExit(
                    "BENCH_RESUME_FROM '{0}': the mode goes after the "
                    "algorithm and before N".format(spec))
            cursor["mode"] = MODES.index(tok)
        else:
            if cursor["algorithm"] is not None or cursor["n"] is not None:
                raise SystemExit(
                    "BENCH_RESUME_FROM '{0}': expected problem[:algorithm]"
                    "[:fixed|adaptive][:N]".format(spec))
            get_algorithm(tok)
            cursor["algorithm"] = algorithm_names().index(tok)
    if cursor["algorithm"] is not None and cursor["mode"] is None:
        cursor["mode"] = 0
    return cursor


def cursor():
    """The parsed BENCH_RESUME_FROM cursor, or None; parsed once."""
    if not _CURSOR_CACHE:
        spec = os.environ.get("BENCH_RESUME_FROM", "")
        _CURSOR_CACHE.append(parse_cursor(spec) if spec else None)
    return _CURSOR_CACHE[0]


def _reset_cache():
    """Forget the parsed cursor (tests change the environment)."""
    del _CURSOR_CACHE[:]


def active():
    """True when any continuation mechanism is switched on."""
    return resume_enabled() or no_overwrite_enabled() or cursor() is not None


def cursor_skips(problem, algorithm, mode, n=None):
    """True when (problem, algorithm, mode[, n]) is before the cursor."""
    cur = cursor()
    if cur is None:
        return False
    pi = problem_names().index(problem)
    if pi != cur["problem"]:
        return pi < cur["problem"]
    if cur["algorithm"] is None:
        return (cur["n"] is not None and n is not None and n < cur["n"])
    ai = algorithm_names().index(algorithm)
    mi = MODES.index(mode)
    if (ai, mi) != (cur["algorithm"], cur["mode"]):
        return (ai, mi) < (cur["algorithm"], cur["mode"])
    return cur["n"] is not None and n is not None and n < cur["n"]


def _status_skips(status):
    """Whether a recorded status is covered under the active flags."""
    if resume_enabled() and status != "absent":
        return True
    return no_overwrite_enabled() and status == "finite"


def skip_point(leg, n, states=None):
    """True when one N or states point of a results.Leg is covered."""
    key = states if leg.analysis == "states" else n
    if cursor_skips(leg.problem.name, leg.algorithm, leg.mode, key):
        return True
    return _status_skips(leg.status(n, states))


def skip_wp_leg(leg, settings):
    """True when every setting of a work-precision results.Leg is covered."""
    if cursor_skips(leg.problem.name, leg.algorithm, leg.mode):
        return True
    if not (resume_enabled() or no_overwrite_enabled()):
        return False
    return all(_status_skips(leg.status(N_WP, setting=setting))
               for setting in settings)


def _cli(argv):
    """Shell entry: "skip"/"run" for a point or a wp leg of a package."""
    from results import Leg
    usage = ("usage: resume.py point <package> <key> <analysis> <problem> "
             "<algorithm> <mode> <N|states> | leg <package> <key> <problem> "
             "<algorithm> <mode>")
    if len(argv) == 8 and argv[0] == "point":
        package, key, analysis, problem, algorithm, mode, value = argv[1:]
        leg = Leg(package, key, analysis, problem, algorithm, mode)
        if analysis == "states":
            skip = skip_point(leg, STATES_N, int(value))
        else:
            skip = skip_point(leg, int(value))
    elif len(argv) == 6 and argv[0] == "leg":
        package, key, problem, algorithm, mode = argv[1:]
        leg = Leg(package, key, "wp", problem, algorithm, mode)
        settings = (leg.problem.dts(algorithm) if mode == "fixed" else TOLS)
        skip = skip_wp_leg(leg, settings)
    else:
        raise SystemExit(usage)
    print("skip" if skip else "run")
    return 0


if __name__ == "__main__":
    import sys
    sys.exit(_cli(sys.argv[1:]))

File: runner_scripts/resume.jl
# Continuation of partial runs against the result store, mirroring resume.py; include problems.jl, algorithms.jl and results.jl first.

const RESUME_MODES = ("fixed", "adaptive")

resume_enabled() = !(get(ENV, "BENCH_RESUME", "") in ("", "0"))

no_overwrite_enabled() = !(get(ENV, "BENCH_NO_OVERWRITE", "") in ("", "0"))

"True when BENCH_FLOOR asks to merge re-runs by keeping the lower time."
floor_enabled() = result_floor_enabled()

"BENCH_RESUME_FROM spec -> cursor tuple; omitted parts are `nothing`."
function parse_resume_cursor(spec)
    parts = String.(split(spec, ':'))
    (isempty(parts) || isempty(parts[1])) &&
        error("BENCH_RESUME_FROM requires a problem name, got '$(spec)'")
    get_problem(parts[1])
    problem = findfirst(==(parts[1]), problem_names())
    algorithm = nothing
    mode = nothing
    n = nothing
    for tok in parts[2:end]
        if occursin(r"^\d+$", tok)
            n === nothing ||
                error("BENCH_RESUME_FROM '$(spec)': more than one N")
            n = parse(Int, tok)
        elseif tok in RESUME_MODES
            (algorithm === nothing || mode !== nothing || n !== nothing) &&
                error("BENCH_RESUME_FROM '$(spec)': the mode goes after " *
                      "the algorithm and before N")
            mode = findfirst(==(tok), collect(RESUME_MODES))
        else
            (algorithm !== nothing || n !== nothing) &&
                error("BENCH_RESUME_FROM '$(spec)': expected " *
                      "problem[:algorithm][:fixed|adaptive][:N]")
            get_algorithm(tok)
            algorithm = findfirst(==(tok), algorithm_names())
        end
    end
    (algorithm !== nothing && mode === nothing) && (mode = 1)
    return (problem = problem, algorithm = algorithm, mode = mode, n = n)
end

const _RESUME_CURSOR = Ref{Any}(:unparsed)

"The parsed BENCH_RESUME_FROM cursor, or `nothing`; parsed once."
function resume_cursor()
    if _RESUME_CURSOR[] === :unparsed
        spec = get(ENV, "BENCH_RESUME_FROM", "")
        _RESUME_CURSOR[] = isempty(spec) ? nothing : parse_resume_cursor(spec)
    end
    return _RESUME_CURSOR[]
end

resume_active() = resume_enabled() || no_overwrite_enabled() ||
                  resume_cursor() !== nothing

"True when (problem, algorithm, mode[, n]) is before the cursor."
function cursor_skips(problem, algorithm, mode, n = nothing)
    cur = resume_cursor()
    cur === nothing && return false
    pi = findfirst(==(problem), problem_names())
    pi == cur.problem || return pi < cur.problem
    if cur.algorithm === nothing
        return cur.n !== nothing && n !== nothing && n < cur.n
    end
    ai = findfirst(==(algorithm), algorithm_names())
    mi = findfirst(==(mode), collect(RESUME_MODES))
    (ai, mi) == (cur.algorithm, cur.mode) ||
        return (ai, mi) < (cur.algorithm, cur.mode)
    return cur.n !== nothing && n !== nothing && n < cur.n
end

"Whether a recorded status is covered under the active flags."
function _status_skips(status)
    resume_enabled() && status != "absent" && return true
    return no_overwrite_enabled() && status == "finite"
end

"True when one N or states point is covered; `key` is N, or the state count in the states sweep."
function skip_point(store, analysis, problem, algorithm, mode, key, n, states,
        setting_kind, setting)
    cursor_skips(problem, algorithm, mode, key) && return true
    status = result_status(store; analysis = analysis, problem = problem,
        algorithm = algorithm, mode = mode, setting_kind = setting_kind,
        setting = _result_fmt(Float64(setting)), n = string(n),
        states = string(states))
    return _status_skips(status)
end

"True when every setting of a work-precision leg is covered."
function skip_wp_leg(store, problem, algorithm, mode, settings, states)
    cursor_skips(problem, algorithm, mode) && return true
    (resume_enabled() || no_overwrite_enabled()) || return false
    kind = mode == "fixed" ? "dt" : "tol"
    return all(settings) do setting
        _status_skips(result_status(store; analysis = "wp", problem = problem,
            algorithm = algorithm, mode = mode, setting_kind = kind,
            setting = _result_fmt(Float64(setting)), n = string(N_WP),
            states = string(states)))
    end
end
